conda_forge_tick/update_sources.py

```diff
@@ -0,0 +1,450 @@
+import abc
+import collections.abc
+import subprocess
+import re
+import logging
+import typing
+from typing import (
+    Any,
+    Optional,
+    Iterable,
+    Set,
+    Iterator,
+    List,
+)
+import yaml
+import feedparser
+import requests
+from conda.models.version import VersionOrder
+from conda_forge_tick.utils import parse_meta_yaml
+from hashing import hash_url
+
+# TODO: parse_version has bad type annotations
+from pkg_resources import parse_version
+
+CRAN_INDEX: Optional[dict] = None
+
+logger = logging.getLogger("conda-forge-tick._update_version.update_sources")
+
+def urls_from_meta(meta_yaml: "MetaYamlTypedDict") -> Set[str]:
+    source: "SourceTypedDict" = meta_yaml["source"]
+    sources: typing.List["SourceTypedDict"]
+    if isinstance(source, collections.abc.Mapping):
+        sources = [source]
+    else:
+        sources = typing.cast("typing.List[SourceTypedDict]", source)
+    urls = set()
+    for s in sources:
+        if "url" in s:
+            # if it is a list for instance
+            if not isinstance(s["url"], str):
+                urls.update(s["url"])
+            else:
+                urls.add(s["url"])
+    return urls
+
+
+def _split_alpha_num(ver: str) -> List[str]:
+    for i, c in enumerate(ver):
+        if c.isalpha():
+            return [ver[0:i], ver[i:]]
+    return [ver]
+
+
+def next_version(ver: str) -> Iterator[str]:
+    ver_split = []
+    ver_dot_split = ver.split(".")
+    n_dot = len(ver_dot_split)
+    for idot, sdot in enumerate(ver_dot_split):
+
+        ver_under_split = sdot.split("_")
+        n_under = len(ver_under_split)
+        for iunder, sunder in enumerate(ver_under_split):
+
+            ver_dash_split = sunder.split("-")
+            n_dash = len(ver_dash_split)
+            for idash, sdash in enumerate(ver_dash_split):
+
+                for el in _split_alpha_num(sdash):
+                    ver_split.append(el)
+
+                if idash < n_dash - 1:
+                    ver_split.append("-")
+
+            if iunder < n_under - 1:
+                ver_split.append("_")
+
+        if idot < n_dot - 1:
+            ver_split.append(".")
+
+    for k in reversed(range(len(ver_split))):
+        try:
+            t = int(ver_split[k])
+        except Exception:
+            continue
+        else:
+            ver_split[k] = str(t + 1)
+            yield "".join(ver_split)
+            ver_split[k] = "0"
```

Comment:
```suggestion
    val_version = re.findall(r"([a-z0-9]+)", ver, re.IGNORECASE)
    separators = re.findall(r"([\-\_\.])", ver)
    re_split_num = re.compile(r"^(\d+)(.*)")

    def join_version(list1, list2):
        result = ""
        for v in chain.from_iterable(zip_longest(list1, list2)):
            if v is None:
                continue
            result += str(v)
        return result

    for i, v in enumerate(reversed(val_version), 1):
        if v.isdecimal():
            val_version[-i] = int(v) + 1
            yield join_version(val_version, separators)
            val_version[-i] = "0"
        else:
            split_num = re_split_num.search(v)
            if not split_num:
                continue
            num_ver, rest_ver = split_num.groups()
            val_version[-i] = f"{int(num_ver) + 1}{rest_ver}"
            yield join_version(val_version, separators)
            val_version[-i] = f"0{rest_ver}"
```

Reply:
Nothing here is simpler. This version obscures what was clear code phrased in terms of loops and simple string operations into calls into more complicated parts of the python API (the `re` module, `chain_from_iterable`, `zip_longest`). 

For newcomers to python, the previous version uses the control structures and API elements they know. The goal is to lower the bar to entry so that more ppl can contribute. This is far more important than saving a few lines of code.

Reply:
```suggestion
    all_versions = re.findall(r"([a-z0-9]+)([\-\_\.]*)?", ver, re.IGNORECASE)
    all_versions = list(x for x in itertools.chain(*all_versions) if x)
    re_split_num = re.compile(r"^(\d+)(.*)")

    for pos, version_number in enumerate(reversed(all_versions), 1):
        if version_number.isdecimal():
            all_versions[-pos] = int(version_number) + 1
            yield all_versions
            all_versions[-pos] = "0"
        else:
            split_num = re_split_num.search(version_number)
            if not split_num:
                continue
            num_ver, rest_ver = split_num.groups()
            all_versions[-pos] = f"{int(num_ver) + 1}{rest_ver}"
            yield all_versions
            all_versions[-pos] = f"0{rest_ver}"
```

Reply:
That I will respectfully disagree with you. Those regex which I used are quite simple, anyone with the very basic of basic of regex would know what is happening there, it is nothing fancy
There are far more regex expressions in ``cf-scripts`` more complex than that
I think it is far more difficult to read 4 nested loops with conditions and all of that, looking for clean code those are considered code smells because they are difficult to read
I even tried to let it more simple
or I can even encapsulate any part in a function and can easily give a better name to it

It also eliminates a couple of functions which will not be necessary, regex does just one good job, which is parsing strings, something which we are trying to do here

Reply:
Ok shrug. Commit what you want.

Reply:
@marcelotrevisani the bot code tends to be a bit finicky, mostly because of lack of test coverage in general (if not in this particular case). Would you be interested in opening an independent PR with the refactors you deem necessary? It would be good to be able to look at them without a large feature being implemented in the same PR.

Reply:
Usually I would say "Sure, I would be happy to do it". But receiving a message such as "**Ok shrug. Commit what you want**" I would prefer to avoid any stress from my side, life is already complicated to stress on opensource projects.

So, I am removing myself from any discussion on ``regro`` in general.

Cheers,

Reply:
My apologies for the tone in that message. I meant only to say that I didn’t want to debate this point anymore. We’d ofc love to have your help here!

Reply:
Folks. There are two schools of though here and both are valid. I'm a scientist who learned to code (:cough: Fortran :cough:), and I appreciate what @beckermr said above. It is close to my coding in general. With that said, @marcelotrevisani is a Computer Scientist from training and has a complete different mind when it comes to coding.

I have to say that, while I resist the change, most of the time I do not regret it. Marcello's code is cleaner, easier to test, maintain, and most of the time, correct on the first try.

BTW, I really dislike regex, mostly b/c I don't understand them and it feels like being left out on a very funny joke. However, than can be easily corrected with comments, different from jokes that lose their meaning when explained, explained regex are a way to include people into a new world.

Reply:
My experience with regex is much the same @ocefpaf and I also dislike them. I don’t find they are well known by new developers and they hide lots of logic in parsing into implicit little bits related to single character changes. 

Again as I said above, please do code with us Marcello!